lib/cretonne/meta/cdsl/ti.py

```diff
@@ -0,0 +1,556 @@
+
+        for (n1, n2, style, elabel) in edges:
+            e = label(n1)
+            if style == "dashed":
+                e += '--'
+            else:
+                e += '->'
+            e += label(n2)
+            e += "[style={}".format(style)
+
+            if elabel is not None:
+                e += ",label={}".format(elabel)
+            e += "];\n"
+
+            r += e
+        r += "}"
+
+        return r
+
+
+if TYPE_CHECKING:
+    TypingError = str
+    TypingOrError = Union[TypeEnv, TypingError]
+
+
+def get_error(typing_or_err):
+    # type: (TypingOrError) -> Optional[TypingError]
+    """
+    Helper function to appease mypy when checking the result of typing.
+    """
+    if isinstance(typing_or_err, str):
+        if (TYPE_CHECKING):
+            return cast(TypingError, typing_or_err)
+        else:
+            return typing_or_err
+    else:
+        return None
+
+
+def get_type_env(typing_or_err):
+    # type: (TypingOrError) -> TypeEnv
+    """
+    Helper function to appease mypy when checking the result of typing.
+    """
+    assert isinstance(typing_or_err, TypeEnv)
+    if (TYPE_CHECKING):
+        return cast(TypeEnv, typing_or_err)
+    else:
+        return typing_or_err
+
+
+def subst(tv, tv_map):
+    # type: (TypeVar, TypeMap) -> TypeVar
+    """
+    Perform substition on the input tv using the TypeMap tv_map.
+    """
+    if tv in tv_map:
+        return tv_map[tv]
+
+    if tv.is_derived:
+        return TypeVar.derived(subst(tv.base, tv_map), tv.derived_func)
+
+    return tv
+
+
+def normalize_tv(tv):
+    # type: (TypeVar) -> TypeVar
+    """
+    Normalize a (potentially derived) TV using the following rules:
+        - collapse SAMEAS
+        SAMEAS(base) -> base
+
+        - vector and width derived functions commute
+        {HALF,DOUBLE}VECTOR({HALF,DOUBLE}WIDTH(base)) ->
+            {HALF,DOUBLE}WIDTH({HALF,DOUBLE}VECTOR(base))
+
+        - half/double pairs collapse
+        {HALF,DOUBLE}WIDTH({DOUBLE,HALF}WIDTH(base)) -> base
+        {HALF,DOUBLE}VECTOR({DOUBLE,HALF}VECTOR(base)) -> base
+    """
+    vector_derives = [TypeVar.HALFVECTOR, TypeVar.DOUBLEVECTOR]
+    width_derives = [TypeVar.HALFWIDTH, TypeVar.DOUBLEWIDTH]
+
+    if not tv.is_derived:
+        return tv
+
+    df = tv.derived_func
+
+    # Collapse SAMEAS edges
+    if (df == TypeVar.SAMEAS):
+        return normalize_tv(tv.base)
+
+    if (tv.base.is_derived):
+        base_df = tv.base.derived_func
+
+        # Reordering: {HALFWIDTH, DOUBLEWIDTH} commute with {HALFVECTOR,
+        # DOUBLEVECTOR}. Arbitrarily pick WIDTH < VECTOR
+        if df in vector_derives and base_df in width_derives:
+            return normalize_tv(
+                    TypeVar.derived(
+                        TypeVar.derived(tv.base.base, df), base_df))
+
+        # Cancelling: HALFWIDTH, DOUBLEWIDTH and HALFVECTOR, DOUBLEVECTOR
+        # cancel each other. TODO: Does this cancellation hide type
+        # overflow/underflow?
```

Comment:
If the domain of the base type variable was restricted at the time the derived type variable is formed, there shouldn't be a problem. When the type functions cancel out, we *don't* increase the base domain again.